Take advantage of C++11, document and clean RS multithreading

G10h4ck opened the pull request:
Please review and test on other platforms like Mac or Windows. According to my research this should work well on all platform (also older), but having it tested would be a plus :) 

A short changelog:
Removed antipattern abstraction RsSingleJobThread
Rename runloop() method to run() in RsThread
Ported few classes ineriting from RsSingleJobThread to RsThread
pqiperson: fix misleading multithreading comment and includes
p3turtle: remove unused forward declaration
RsThread fix includes
Properly enable C++11 both for Qt4 and Qt5
RsThread use std::atomic<bool> (supported since gcc-4.4) instead of self implemented strange binary semaphores
Removed RsTickingThread::shutdown() use RsThread::askForStop() instead
Removed RsTickingThread::fullstop() use RsThread::fullstop() instead

PhenomRetroShare committed:
Fix El Capitan OSX 10.11 Compil

G10h4ck committed:
Enable C++11 and document and clean multithreading

Removed antipattern abstraction RsSingleJobThread
Rename runloop() method to run() in RsThread
Ported few classes ineriting from RsSingleJobThread to RsThread
pqiperson: fix misleading multithreading comment and includes
p3turtle: remove unused forward declaration
RsThread fix includes
Properly enable C++11 both for Qt4 and Qt5
RsThread use std::atomic<bool> (supported since gcc-4.4) instead of self implemented strange binary semaphores

G10h4ck committed:
More restructuring of RS threading

Removed RsTickingThread::shutdown() use RsThread::askForStop() instead
Removed RsTickingThread::fullstop() use RsThread::fullstop() instead
More documentation for RsThread and RsTickingThread

G10h4ck commented:
@PhenomRetroShare a bird said to me that you could test this on Mac, would you please? We are actually waiting for that kind of testing before merging this

PhenomRetroShare commented:
Can you update your branch I have to merge this one on it to compil:
https://github.com/PhenomRetroShare/RetroShare/tree/Fix_ElCapitanCompil
This branch works fine on 10.9, but crash on 10.11.
Or I can make a PR with mine and you rebase on it.

G10h4ck commented:
@PhenomRetroShare just pushd this branch
https://github.com/G10h4ck/RetroShare/tree/rsthread_cleanup_ElCapitanCompil

I have rebased on top of the branch you proposed, may you test it ?

Thanks!

PhenomRetroShare commented:
OK, this compils and runs.
But don't test connection as test node.

On Sun, Sep 18, 2016 at 7:31 PM, Gioacchino notifications@github.com
wrote:

> @PhenomRetroShare https://github.com/PhenomRetroShare just pushd this
> branch
> https://github.com/G10h4ck/RetroShare/tree/rsthread_
> cleanup_ElCapitanCompil
> 
> I have rebased on top of the branch you proposed, may you test it ?
> 
> Thanks!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/RetroShare/RetroShare/pull/495#issuecomment-247861207,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AI-IfTqOtYuNXyyd-j60kzlxY7f-oJYEks5qrXVogaJpZM4J664k
> .

G10h4ck commented:
@PhenomRetroShare what you mean by "But don't test connection as test node." is that you haven't tested it yet, or that it doesn't worked?

PhenomRetroShare commented:
My Mac test node has no friend. So I can't test connection between nodes.

On Sun, Sep 18, 2016 at 9:14 PM, Gioacchino notifications@github.com
wrote:

> @PhenomRetroShare https://github.com/PhenomRetroShare what you mean by
> "But don't test connection as test node." is that you haven't tested it
> yet, or that it doesn't worked?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/RetroShare/RetroShare/pull/495#issuecomment-247867276,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AI-Ifao4s8P6UvtWhP0DZW_PNC4V6zPBks5qrY2ZgaJpZM4J664k
> .

G10h4ck commented:
This pull request should not alter connection behaviour, but if you wanna be sure we can add your node as friend somewhere (for example Just Relay It)

PhenomRetroShare commented:
It's was running 1H30 with 15 friend. Nothing to report.
But not running under debugger neither terminal, so no output.
I can test test it more time this weekend if needed.

PhenomRetroShare commented:
Some changes to get RS re-based and compiled under Windows:

```
--- libretroshare/src/file_sharing/directory_updater.cc
+++ libretroshare/src/file_sharing/directory_updater.cc
@@ -57,7 +57,7 @@ void LocalDirectoryUpdater::setEnabled(bool b)
     if(b)
         start() ;
     else
-        shutdown();
+        askForStop();

     mIsEnabled = b ;
 }
--- libretroshare/src/file_sharing/hash_cache.cc
+++ libretroshare/src/file_sharing/hash_cache.cc
@@ -115,7 +115,7 @@ void HashStorage::data_tick()
                 if(!mChanged)  // otherwise it might prevent from saving the hash cache
                 {
                     std::cerr << "Stopping hashing thread." << std::endl;
-                    shutdown();
+                    askForStop();
                     mRunning = false ;
                     mTotalSizeToHash = 0;
                     mTotalFilesToHash = 0;
--- libretroshare/src/pqi/pqiperson.h
+++ libretroshare/src/pqi/pqiperson.h
@@ -25,13 +25,17 @@

 #pragma once

-#include <stdint.h>              // for uint32_t, uint16_t
-#include <sys/socket.h>          // for sockaddr_storage
-#include <time.h>                // for NULL, time_t
 #include <iostream>              // for operator<<, basic_ostream, cerr, endl
 #include <list>                  // for list
 #include <map>                   // for map
+#include <stdint.h>              // for uint32_t, uint16_t
 #include <string>                // for string, char_traits
+#include <time.h>                // for NULL, time_t
+#ifndef WINDOWS_SYS
+#include <sys/socket.h>          // for sockaddr_storage
+#else
+#include <winsock2.h>            // for sockaddr_storage
+#endif

 #include "pqi/pqi_base.h"        // for NetBinInterface, NetInterface, PQInt...
 #include "pqi/pqiqosstreamer.h"  // for pqiQoSstreamer
```

Could you update this branch?
And does someone get it working on windows?
I've got this:

```
Thread 4 (Thread 1228.0xcc4):
#0  0x778a6bf4 in ntdll!KiFastSystemCallRet () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#1  0x778a656c in ntdll!ZwWaitForMultipleObjects () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#2  0x7788fcdf in ntdll!RtlFreeThreadActivationContextStack () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#3  0x00000003 in ?? ()
No symbol table info available.
#4  0x0027d0e0 in ?? ()
No symbol table info available.
#5  0x773cef1c in KERNEL32!BaseThreadInitThunk () from C:\Windows\system32\kernel32.dll
No symbol table info available.
#6  0x778c3648 in ntdll!RtlInitializeExceptionChain () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#7  0x778c361b in ntdll!RtlInitializeExceptionChain () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#8  0x00000000 in ?? ()
No symbol table info available.

Thread 3 (Thread 1228.0x4c0):
#0  0x778a6bf4 in ntdll!KiFastSystemCallRet () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#1  0x778a659c in ntdll!ZwWaitForWorkViaWorkerFactory () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#2  0x778910f5 in ntdll!TpDisassociateCallback () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#3  0x773cef1c in KERNEL32!BaseThreadInitThunk () from C:\Windows\system32\kernel32.dll
No symbol table info available.
#4  0x778c3648 in ntdll!RtlInitializeExceptionChain () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#5  0x778c361b in ntdll!RtlInitializeExceptionChain () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#6  0x00000000 in ?? ()
No symbol table info available.

Thread 1 (Thread 1228.0x78c):
#0  0x00aa2e37 in QString::QString (this=0x400000, dd=...) at C:/Qt/5.5/mingw492_32/include/QtCore/qstring.h:765
No locals.
#1  0x00cf91ce in Ui_IdDetailsDialog::setupUi(QDialog*)::{lambda()#16}::operator()() const (__closure=0x400000) at temp/ui/ui_IdDetailsDialog.h:139
        qstring_literal = {str = {ref = {atomic = {_q_value = -1}}, size = 14, alloc = 0, capacityReserved = 0, offset = 16, static shared_null = {{ref = {atomic = {_q_value = -1}}, size = 0, alloc = 0, capacityReserved = 0, offset = 16, static shared_null = <same as static member of an already seen type>}, {ref = {atomic = {_q_value = 0}}, size = 0, alloc = 0, capacityReserved = 0, offset = 0, static shared_null = <same as static member of an already seen type>}}}, data = u"lineEdit_GpgId"}
        holder = {ptr = 0x148d380 <Ui_IdDetailsDialog::setupUi(QDialog*)::{lambda()#16}::operator()() const::qstring_literal>}
        qstring_literal_temp = {static null = {<No data fields>}, d = 0x905a4d <resource_api::JsonStream::setError()+21>}
#2  0x6337cd39 in ?? () from C:\msys32\mingw32\bin\libeay32.dll
No symbol table info available.
#3  0x633009ea in ?? () from C:\msys32\mingw32\bin\libeay32.dll
No symbol table info available.
#4  0x632ffe6d in ?? () from C:\msys32\mingw32\bin\libeay32.dll
No symbol table info available.
#5  0x006ba7f6 in RSRandom::locked_next_state () at ..\..\..\RetroShare\libretroshare\src\util\rsrandom.cc:50
No locals.
#6  0x006ba77e in RSRandom::seed (s=1089275290) at ..\..\..\RetroShare\libretroshare\src\util\rsrandom.cc:39
        mtx = {mMtx = @0x185f16c}
        j = 1024
#7  0x006bab18 in __static_initialization_and_destruction_0 (__initialize_p=1, __priority=65535) at ..\..\..\RetroShare\libretroshare\src\util\rsrandom.cc:21
No locals.
#8  0x006bab78 in _GLOBAL__sub_I__ZN8RSRandom5indexE () at ..\..\..\RetroShare\libretroshare\src\util\rsrandom.cc:101
No locals.
#9  0x0090eaba in __do_global_ctors ()
No symbol table info available.
#10 0x004013b2 in __tmainCRTStartup ()
No symbol table info available.
#11 0x773cef1c in KERNEL32!BaseThreadInitThunk () from C:\Windows\system32\kernel32.dll
No symbol table info available.
#12 0x778c3648 in ntdll!RtlInitializeExceptionChain () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#13 0x778c361b in ntdll!RtlInitializeExceptionChain () from C:\Windows\SYSTEM32\ntdll.dll
No symbol table info available.
#14 0x00000000 in ?? ()
No symbol table info available.
```

G10h4ck committed:
fix compilation with new RsThread API

PhenomRetroShare commented on libretroshare/src/file_sharing/hash_cache.cc:
```diff
@@ -117,8 +117,8 @@ void HashStorage::data_tick()
                     std::cerr << "Stopping hashing thread." << std::endl;
-                    shutdown();
-                    mRunning = false ;
+					askForStop();
```
Take care of indent character.
Show your white-space.

sehraf replied:
@G10h4ck is just following the coding style. Everybody else should do the same!
https://github.com/RetroShare/RetroShare/wiki/Coding#coding-style

PhenomRetroShare replied:
Yes I'm 100% agree, but this is preferable to follow the style when a file is already wrote.

sehraf replied:
I don't agree - but this comes down to a matter of taste. Remove two tabs and it almost looks fine in the commit again 😸

PhenomRetroShare commented:
You miss this for windows compilation:

```
--- libretroshare/src/pqi/pqiperson.h
+++ libretroshare/src/pqi/pqiperson.h
@@ -25,13 +25,17 @@

 #pragma once

-#include <stdint.h>              // for uint32_t, uint16_t
-#include <sys/socket.h>          // for sockaddr_storage
-#include <time.h>                // for NULL, time_t
 #include <iostream>              // for operator<<, basic_ostream, cerr, endl
 #include <list>                  // for list
 #include <map>                   // for map
+#include <stdint.h>              // for uint32_t, uint16_t
 #include <string>                // for string, char_traits
+#include <time.h>                // for NULL, time_t
+#ifndef WINDOWS_SYS
+#include <sys/socket.h>          // for sockaddr_storage
+#else
+#include <winsock2.h>            // for sockaddr_storage
+#endif

 #include "pqi/pqi_base.h"        // for NetBinInterface, NetInterface, PQInt...
 #include "pqi/pqiqosstreamer.h"  // for pqiQoSstreamer
```

But always this SIGSEGV, at a new place:

```
1  QString::QString                                                          qstring.h            765 0xaa371b   
2  Ui_GxsForumMsgItem::setupUi(QWidget *)::{lambda()#23}::operator()() const ui_GxsForumMsgItem.h 212 0xcf96aa   
3  Ui_GxsForumMsgItem::setupUi(QWidget *)::{lambda()#23}::_FUN()             ui_GxsForumMsgItem.h 212 0xcf9f23   
4  ??                                                                                                 0x6337cde9 
5  ??                                                                                                 0x633009fa 
6  ??                                                                                                 0x632ffe7d 
7  RSRandom::locked_next_state                                               rsrandom.cc          50  0x6ba8b6   
8  RSRandom::seed                                                            rsrandom.cc          39  0x6ba83e   
9  __static_initialization_and_destruction_0                                 rsrandom.cc          21  0x6babd8   
10 _GLOBAL__sub_I__ZN8RSRandom5indexE                                        rsrandom.cc          101 0x6bac38   
11 __do_global_ctors                                                                                  0x90f33a   
12 __tmainCRTStartup                                                                                  0x4013b2   
13 KERNEL32!BaseThreadInitThunk                                                                       0x7764ef1c 
14 ntdll!RtlInitializeExceptionChain                                                                  0x77d03648 
15 ntdll!RtlInitializeExceptionChain                                                                  0x77d0361b 
16 ??                                                                                                            
```

I'm looking on that importing modification step by step.

PhenomRetroShare commented:
Found.
It's CONFIG += c++11
Don't know why.
Changing form QT5.5 to QT.5.7.
Edit:
It's the same.

G10h4ck commented:
What do you mean?

Is CONFIG += c++11 responsible of the crash or it is that CONFIG += c++11 is missing in windows parts of .pro files?

G10h4ck committed:
Fix compilation on windows

sehraf commented:
sorry for nagging again about this but if #359 would be merged (or maybe redone) we could manage configs like c++11 in one central place for everything

G10h4ck commented:
Last conversation i had with phenom he explained that CONFIG += C++11 seems responsible for the crash on windows.

If someone help me setting up a windows compiling environment i could investigate this further.

Any news with this?

PhenomRetroShare commented:
@G10h4ck : You have this file to help you:
https://github.com/RetroShare/RetroShare/blob/master/WindowsMSys2_InstallGuide.txt

G10h4ck commented:
From a discussion we are having in the chat lobby seems the crash is not specific of this pull request but it happens on master too, I have copied the discussion on a forum thread to keep memory of it, so as far as I understand the crash is more specific of the way how the binary is built, what we need is that the ones that build RetroShare installers for windows write the documentation on how to do it correctly, so i would go forward with this merge request and merge it.

retroshare://forum?name=Re%3A Seems that RS crash on Win7&id=95de1451952d8c38cb1cdfdb85eed986&msgid=877d0c2c1ff090949301d8557c7b0b2f8ca265b9

@csoler any comment?

PhenomRetroShare commented:
Build master with Qt5.5.1 works fine.
But adding CONFIG += c++11 crash at the same place than with Qt5.7.
So you PR crash on Qt5.5.1 too.

G10h4ck commented:
From discussion I had with #qt folks they say it may be related to some static initialization order stuff, so potentially the code is broken on everyplatform, but the crash is actually triggering only on windows + C++11

I have been looking at RsRandom code and the only thing that may be "suspect" is from like 12 to 25
https://github.com/RetroShare/RetroShare/blob/master/libretroshare/src/util/rsrandom.cc#L12

@PhenomRetroShare would you try testing removing those line and see what happens?

G10h4ck commented:
more info in the RS forum

retroshare://forum?name=Re%3A Seems that RS crash on Win7 C%2B%2B11&id=95de1451952d8c38cb1cdfdb85eed986&msgid=967525b2333096d6b56c05cb666cf660196fbc6a

PhenomRetroShare commented:
This move the crash:
1   QString::QString                                                       qstring.h               791  0xac4f6f  
2   Ui_TransferPage::setupUi(QWidget *)::{lambda()#14}::operator()() const ui_TransferPage.h       103  0xd21146  
3   ??                                                                                                  0x6337cde9 
4   ??                                                                                                  0x633009fa 
5   ??                                                                                                  0x632ffe7d 
6   ??                                                                                                  0x632ba4f8 
7   ??                                                                                                  0x632bd061 
8   ??                                                                                                  0x632bdacb 
9   ??                                                                                                  0x632da158 
10  ops_rsa_generate_keypair                                               openssl_crypto.c        674  0x8f55f8  
11  ops_rsa_create_selfsigned_keypair                                      openssl_crypto.c        769  0x8f58a6  
12  PGPHandler::GeneratePGPCertificate                                     pgphandler.cc           414  0x77b67a  
13  AuthGPG::GeneratePGPCertificate                                        authgpg.cc              343  0x7876d3  
14  RsAccountsDetail::GeneratePGPCertificate                               rsaccounts.cc           896  0x72b8cd  
15  RsAccounts::GeneratePGPCertificate                                     rsaccounts.cc           1275 0x72d0d9  
16  GenCertDialog::genPerson                                               GenCertDialog.cpp       597  0x41b0fa  
17  GenCertDialog::qt_static_metacall                                      moc_GenCertDialog.cpp   86   0x65dce1  
18  QMetaObject::activate                                                  qobject.cpp             3740 0x18e056da 
19  QMetaObject::activate                                                  qobject.cpp             3602 0x18e04ffa 
20  QAbstractButton::clicked                                               moc_qabstractbutton.cpp 307  0x32545571 
... <plus>

G10h4ck commented:
I have an hypothesis, Q_DECL_CONSTEXPR behave differently with C++11 and C++98, this probably cause that Qt compiled with C++98 are ABI incompatible with Qt compiled with C++11 al least on windows (probably unixex can handle multiple ABIs), so when we compile RetroShare with CONFIG+=C++11 with a precompiled Qt that wasn#t compiled with C++11 the method of QString that is declared as Q_DECL_CONSTEXPR cause crash and probably stack corruption too (as the backtrace make no sense at all). As you know compiling on windows is very slow so to verify this hypothesis will require some time as i have to compile the whole Qt and RetroShare again, I hope to come back with good news soon.

EDIT: This proved to be false!

G10h4ck committed:
Substitute non crossplatform Sleep with rs_usleep

While doing this i have discovered a stacked mutex in the same block of rs_usleep(...) in
RsControlModule::askForPassword, dixed this by moving the stacked mutex
and the critical code inside a block

G10h4ck committed:
Fix nasty crash on windows due to RsRandom

In C++ static initialization order is not guaranted outside the same
compilation unit, interdependet static initialization is a bad practice
and can lead to unpredictable behaviour higly depending on compiler and
platform, RsRandom::N was used as parameter by other members
initializations, in particular as an std::vector size thus corrupting
the memory and causing a crash at init. This problem has been solved by
storing the RsRandom state size in a macro local to rsrandom.cc and use
it instead of RsRandom::N that has been completely removed.

RsRandom::random_alphaNumericString(...) doesn't pass the result string
by value anymore, instead it take a reference where the result is
stored.

RSRandom first time seeding has been moved inside
RsRandom::random_bytes(...) and guarded by a static boolean so it is
executed just first time, moreover it is now enabled on all platforms
not just windows, coerently to the provided code comment.

G10h4ck committed:
Merge branch 'master' into rsthread_cleanup

G10h4ck committed:
Merge branch 'master' into rsthread_cleanup

G10h4ck committed:
Furter cleanup of RsRandom

Updated documentation (RsRandom is crypto safe)
Removed unuseful resize on random state vector

G10h4ck committed:
Tentative to avoid crash on AuthPGP in windows

G10h4ck committed:
Tentative avoid crash in PGP key generation

G10h4ck committed:
Get rid of the orrorous rsAccounts extern global pointer

G10h4ck committed:
Reverted porcata code introduced while investigating crash

G10h4ck committed:
Some cleanup in pgphandler and authgpg

dtschmitz commented:
Why has this PR been delayed?  It is of paramount importance.  You have major segfaults from pthread issues.  Do it without further delay @csoler and it should NOT be tagged as a 'Feature Enhancement'.  It's a compiler switch.  cc: @PhenomRetroShare @G10h4ck @sehraf @cavebeat

dtschmitz commented:
'''
 @cavebeat added the enhancement label [on Sep 19, 2016](https://github.com/RetroShare/RetroShare/pull/495#event-794527750)
'''
This is stupid.  Why don't you add every feature in C++ as an enhancement?  
It's not an enhancement to RS functionality.  It is a compiler switch.
cc: @csoler @G10h4ck @PhenomRetroShare

G10h4ck commented:
@dtschmitz please avoid verbal violence usage as it is not necessary, this merge request is outdated and it is not closed just to remember a bunch of improvements that should be implemented in a proper way in the future (for sure not before 0.6.3)

G10h4ck commented:
Superseded by newer and more specific https://github.com/RetroShare/RetroShare/pull/1698
